cmd/snapd-generator/main.c

```diff
@@ -97,3 +91,106 @@ int main(int argc, char **argv)
 	fprintf(f, "WantedBy=local-fs.target\n");
 	return 0;
 }
+
+static bool file_exists(const char *path) {
+	struct stat buf;
+	return stat(path, &buf) == 0;
+}
+
+static bool executable_exists(const char *name) {
+	char *path = getenv("PATH");
+	if (path == NULL) {
+		return false;
+	}
+	char *path_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+	path_copy = sc_strdup(path);
+
+	char *ptr = NULL;
+	char *token = strtok_r(path_copy, ":", &ptr);
+	char fname[PATH_MAX + 1] = { 0 };
+	while (token) {
+		sc_must_snprintf(fname, sizeof fname, "%s/%s", token, name);
```

Comment:
Non-binding comment: another way to do this without string manipulation and running into weird length limits is to `open(2)` each component via O_PATH|O_CLOEXEC|O_DIRECTORY, and to `fstatat(2)`.